File: Design2_Problem1.py
# Time Complexity : O(1) for push, peek, O(n) for pop operation
# Space Complexity : O(n)
# Did this code successfully run on Leetcode : yes
# Any problem you faced while coding this : None

# Your code here along with comments explaining your approach : using 2 stacks I pop from stack 1 and add to stack 2 and check both the stacks for the output

# BruteForce 2
class MyQueue(object):

    def __init__(self):
        self.stack = []
        self.opp_stack = []

    def push(self, x):
        """
        :type x: int
        :rtype: None
        """
        self.stack.append(x)
        

    def pop(self):
        """
        :rtype: int
        """ 
        if self.opp_stack:
            return self.opp_stack.pop()
        while self.stack:
            self.opp_stack.append(self.stack.pop())
        temp = self.opp_stack.pop()
        # while self.opp_stack:
        #     self.stack.append(self.opp_stack.pop())
        return temp
        

    def peek(self):
        """
        :rtype: int
        """
        return self.stack[0] if bool(self.stack) else self.opp_stack[-1]

    def empty(self):
        """
        :rtype: bool
        """
        return not bool(self.stack) and not bool(self.opp_stack)


# Your MyQueue object will be instantiated and called as such:
# obj = MyQueue()
# obj.push(x)
# param_2 = obj.pop()
# param_3 = obj.peek()
# param_4 = obj.empty()



# BruteForce
# class MyQueue(object):

#     def __init__(self):
#         self.stack = []
#         self.opp_stack = []

#     def push(self, x):
#         """
#         :type x: int
#         :rtype: None
#         """
#         self.stack.append(x)
        

#     def pop(self):
#         """
#         :rtype: int
#         """ 
#         if len(self.stack) == 0 and len(self.opp_stack) == 0:
#             return "Q is empty"
#         elif len(self.opp_stack) == 0:
#             while self.stack:
#                 self.opp_stack.append(self.stack.pop())
#             return self.opp_stack.pop()
#         else:
#             return self.opp_stack.pop()

#     def peek(self):
#         """
#         :rtype: int
#         """
#         return self.opp_stack[-1] if self.opp_stack else self.stack[0]

#     def empty(self):
#         """
#         :rtype: bool
#         """
#         return not bool(self.stack) and not bool(self.opp_stack)

File: Design2_Problem1.java
// Time Complexity : Push : O(1), Pop and Peek : O(N)

// Space Complexity : O(N)

// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach


/*

Created 2 Stacks s1 anbd s2
In push : PUsh elements in s1

In Pop : Pop all elements from s1 and push it into s2
Now pop the top element from s2 and return it // Deleted element from front of the queue
Push back all elements from s2 to s1


In peek :  Pop all elements from s1 and push it into s2
Now peek the top element from s2 and return it // element at front of the queue
Push back all elements from s2 to s1

*/
import java.util.Stack;

class Mys2 {
    Stack<Integer> s2,s1;
    public Mys2() {
        s1 = new Stack<Integer>();
        s2 = new Stack<Integer>();
    }
    public void push(int x) {
        s1.push(x);  
    }
    public int pop() {
        while(! s1.isEmpty()){
            s2.push(s1.pop());
        }
        int element = s2.pop(); //Top-most element deleted
        
        while(! s2.isEmpty() ){
            s1.push(s2.pop());
        }
        return element;
    }
    
    public int peek() {
        while(! s1.isEmpty()){
            s2.push(s1.pop());
        }
        int top = s2.peek(); //Top-most element selected
        
        while(! s2.isEmpty() ){
            s1.push(s2.pop());
        }    
        return top;
    }
    
    public boolean empty() {
        return s1.isEmpty();
    }
}

public class Design2_Problem1 {
    public static void main(String[] args){
        Mys2 s2 = new Mys2();
        s2.push(1);  s2.push(2); 
    
        System.out.println(("s2.peek() : " + s2.peek())); // FRONY ELEMENT

        System.out.println("s2.empty() : " + s2.empty() );    
    }


}
